### PR TITLE
Switch to Xxhash128 from sha256

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes [AB#1880936](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1880936)

xxhash128 is smaller, and *much faster* (around 40x so) while still retaining the collision resistance we need for non-cryptographic content hashing.  The API is also much easier to use, resulting in a ton of complex netfx vs netcore code being deleted.

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Solution/Checksum.cs [hunk 1]
```diff
@@ -119,29 +120,24 @@ public override int GetHashCode()
...
             {
-                Contract.ThrowIfFalse(span.Length >= HashSize);
+                Contract.ThrowIfTrue(span.Length < HashSize);
 #pragma warning disable CS9191 // The 'ref' modifier for an argument corresponding to 'in' parameter is equivalent to 'in'. Consider using 'in' instead.
                 Contract.ThrowIfFalse(MemoryMarshal.TryWrite(span, ref Unsafe.AsRef(in this)));
 #pragma warning restore CS9191
             }
 
-            public static unsafe HashData FromPointer(HashData* hash)
-                => new(hash->Data1, hash->Data2, hash->Data3);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/Checksum.cs:141]: never used.

File: src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs [hunk 2]
```diff
@@ -21,135 +19,41 @@ namespace Microsoft.CodeAnalysis
     internal partial record class Checksum
     {
         // https://github.com/dotnet/runtime/blob/f2db6d6093c54e5eeb9db2d8dcbe15b2db92ad8c/src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/SHA256.cs#L18-L19
-        private const int SHA256HashSizeBytes = 256 / 8;
+        private const int XXHash128SizeBytes = 128 / 8;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs:22]: not sure why we need this constant  *and* the HashSize constant.  But was keeping this roughly the same.

File: src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs [hunk 3]
```diff
@@ -21,135 +19,41 @@ namespace Microsoft.CodeAnalysis
...
-            new(SHA256.Create, size: 20);
-#endif
-
-#if !NET
-        // Dedicated pools for the byte[]s we use to create checksums from two or three existing checksums. Sized to
-        // exactly the space needed to splat the existing checksum data into the array and then hash it.
-
-        // Note: number of elements here should be computed based on what we need from our various collection-with-children objects.
-        private static readonly ObjectPool<byte[]>[] s_checksumByteArrayPool =
-            Enumerable.Range(0, 11).Select(i => new ObjectPool<byte[]>(() => new byte[HashSize * i])).ToArray();
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs:40]: no longer needed.  the xxhash128 has all the functionality we need.

File: src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs [hunk 4]
```diff
@@ -21,135 +19,41 @@ namespace Microsoft.CodeAnalysis
...
         public static Checksum Create(Stream stream)
         {
-#if NET7_0_OR_GREATER
-            Span<byte> hash = stackalloc byte[SHA256HashSizeBytes];
-            SHA256.HashData(stream, hash);
-            return From(hash);
-#elif NET
             using var pooledHash = s_incrementalHashPool.GetPooledObject();
-            Span<byte> buffer = stackalloc byte[SharedPools.ByteBufferSize];
+            pooledHash.Object.Append(stream);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs:52]: xxhash128 already has support for appending a stream to it.  woohoo.

File: src/Workspaces/CoreTest/ChecksumTests.cs [hunk 5]
```diff
@@ -19,7 +19,7 @@ public void ValidateChecksumFromSpanSameAsChecksumFromBytes1()
             var checksumA = Checksum.Create(checksum1, checksum2);
 
             // Running this test on multiple target frameworks with the same expectation ensures the results match
-            Assert.Equal(Checksum.FromBase64String("N30m5jwVeMZzWpy9cbQbtSYHoXU="), checksumA);
+            Assert.Equal(Checksum.FromBase64String("RRbspG+E4ziBC5hOWyrfCQ=="), checksumA);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CoreTest/ChecksumTests.cs:22]: expected change.  we are legitimately using a different hash.  

the tests *do* check though that we get the same result on netfx vs netcore (that way we can have roslyn communicating across process while still agreeing on the results).

**Comment by sharwell** [comment]
@DustinCampbell We should probably switch Razor at the same time if we are making this change

File: src/Workspaces/Core/Portable/Workspace/Solution/Checksum.cs [hunk 6]
```diff
@@ -13,6 +13,7 @@
 using System.Runtime.Serialization;
 using Microsoft.CodeAnalysis.Serialization;
 using Roslyn.Utilities;
+using System.IO.Hashing;
```
Comment (tmat) [inline src/Workspaces/Core/Portable/Workspace/Solution/Checksum.cs:16]: >System [](http://example.com/codeflow?start=6&length=6)

nit: sort

File: src/Workspaces/Core/Portable/Workspace/Solution/Checksum.cs [hunk 7]
```diff
@@ -119,29 +120,24 @@ public override int GetHashCode()
         [DataContract, StructLayout(LayoutKind.Explicit, Size = HashSize)]
         public readonly record struct HashData(
```
Comment (tmat) [inline src/Workspaces/Core/Portable/Workspace/Solution/Checksum.cs:121]: >HashData [](http://example.com/codeflow?start=38&length=8)

Can we use InlineArray feature instead?

File: src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs (hunk 2 above)
Comment (tmat) [reply src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs:22]: Could you follow up with cleanup-refactoring? 
Other than having this duplicated I'm not sure about the value of having the Create methods in a separate file.

File: src/Workspaces/Core/Portable/Workspace/Solution/Checksum.cs (hunk 7 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/Checksum.cs:120]: no.  this has to run on netfx and core.

File: src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs:22]: yes.  i'm also intending to measure why we need Checksum *and* HashData when it seems like we could just have the latter.  The former is technically good if we expect a few actual distinct checksums, and tons of pointers to them.  The latter would be good if we expect less pointers and most of the values to just be needed in the place where we're pointing at them.

File: eng/Versions.props [hunk 8]
```diff
@@ -245,6 +245,7 @@
     <SystemDrawingCommonVersion>7.0.0</SystemDrawingCommonVersion>
     <SystemIOFileSystemVersion>4.3.0</SystemIOFileSystemVersion>
     <SystemIOFileSystemPrimitivesVersion>4.3.0</SystemIOFileSystemPrimitivesVersion>
+    <SystemIOHashingVersion>8.0.0-rc.2.23479.6</SystemIOHashingVersion>
```
Comment (AArnott) [inline eng/Versions.props:248]: We won't want to ship a prerelease of this dependency with VS. Can you use the latest stable version instead?

Comment (CyrusNajmabadi) [reply eng/Versions.props:248]: we cannot.   the latest stable does not include xxhash128 (the actual hashing function we're trying to use).

**Comment by CyrusNajmabadi** [comment]
Going to hold off on this till next tuesday when 8.0 releases.

File: src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs (hunk 2 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs:22]: We might have been a 32-bit process the last time this was evaluated, but the concern last time we looked was the number of pointers could exceed the number of instances.

File: eng/Versions.props (hunk 8 above)
Comment (CyrusNajmabadi) [inline eng/Versions.props]:
```suggestion
    <SystemIOHashingVersion>8.0.0</SystemIOHashingVersion>
```

**Comment by CyrusNajmabadi** [comment]
Note: this will need changes to VS build to pull in System.IO.Hashing. @jasonmalinowski are you a good person to coordinate that with?

**Review by jasonmalinowski — APPROVED** [review]

File: src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs [hunk 9]
```diff
@@ -21,135 +19,41 @@ namespace Microsoft.CodeAnalysis
     internal partial record class Checksum
     {
         // https://github.com/dotnet/runtime/blob/f2db6d6093c54e5eeb9db2d8dcbe15b2db92ad8c/src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/SHA256.cs#L18-L19
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs:21]: sure :)

**Comment by CyrusNajmabadi** [comment]
@jaredpar @jasonmalinowski the source-build change went in with: https://github.com/dotnet/source-build-reference-packages/pull/822

Do you know why this is still having issues building on our side?  Thanks! :)

**Comment by lbussell** [comment]
You may need to pick up the latest version of SBRP - similar to https://github.com/dotnet/roslyn/pull/69901.

**Comment by MichaelSimons** [comment]
> You may need to pick up the latest version of SBRP - similar to #69901.

Ideally long term a darc subscription would be setup to automatically flow new versions in.

**Pull request merged**